BUG: Plotting use FixedLocator

nrebena opened the pull request:
For string x values or multiindex, the axis locator is now set to a
FixedLocator for Line and Area plot. The ticklabel stay consistent
when scaling and zooming.

closes #7612 
closes #15912
closes #22334
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

nrebena committed:
BUG: Plotting use FixedLocator

For string x values or multiindex, the axis locator is now set to a
FixedLocator for Line and Area plot. The ticklabel stay consistent
when scaling and zooming.
This impact the following issues:
GH7612, GH15912, GH22334

WillAyd commented:
Please add tests to cover the issues being closed

Commit:
TST: Add test for #7612 #22334 #15912

pep8speaks commented:
Hello @nrebena! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/plotting/test_frame.py`](https://github.com/pandas-dev/pandas/blob/7cea50ae6de4b359d9df57c8819296b977ba7d40/pandas/tests/plotting/test_frame.py):

> [Line 3004:20](https://github.com/pandas-dev/pandas/blob/7cea50ae6de4b359d9df57c8819296b977ba7d40/pandas/tests/plotting/test_frame.py#L3004): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 3005:20](https://github.com/pandas-dev/pandas/blob/7cea50ae6de4b359d9df57c8819296b977ba7d40/pandas/tests/plotting/test_frame.py#L3005): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

Commit:
CLN: Fiex visual indent for PEP8

nrebena commented:
I added tests for the issue. I used the example from #22334.
```
        df = pd.DataFrame({'sales': [3, 2, 3],
                           'visits': [20, 42, 28], 
                           'day': ['Monday', 'Tuesday', 'Wednesday']})
        ax = df.plot.area(x='day')
        ax.set_xlim(-1, 3)
        xticklabels = [t.get_text() for t in ax.get_xticklabels()]
        labels_position = dict(zip(xticklabels, ax.get_xticks()))
        # Testing if the label stayed at the right position
        assert labels_position['Monday'] == 0.0
        assert labels_position['Tuesday'] == 1.0
        assert labels_position['Wednesday'] == 2.0
```
![Figure_1](https://user-images.githubusercontent.com/49879400/56582059-aceb0180-65d6-11e9-81e6-c5abb4f952cb.png)

This should guarantee that the xticklabel stay at the same position even after a change in the axis.
I am open for idea if there is a better way to test for the issues.

jreback commented on pandas/plotting/_core.py:
```diff
@@ -410,8 +410,13 @@ def get_label(i):
+                xticks = ax.get_xticks()
+                xticklabels = [get_label(x) for x in xticks]
                 ax.set_xticklabels(xticklabels)
+                from matplotlib.ticker import FixedLocator, FixedFormatter
```
can you put imports higher up in this function to avoid need to import in separate locations

jreback commented on pandas/tests/plotting/test_frame.py:
```diff
@@ -2995,6 +2995,23 @@ def test_secondary_axis_font_size(self, method):
 
+    @pytest.mark.slow
+    def test_x_string_values_ticks(self):
+        # GH: 7612
```
can you add a 1-liner explaining what this test is doing; all of the issue references can be on a single line

nrebena replied:
I added the following one-liner:
# Test if string plot index have a fixed xtick position
# Test if multiindex plot index have a fixed xtick position

jreback reviewed (changes requested):
pls add a whatsnew note, can be a single one for all 3 issues if they are actually the same.
pls make sure that all of the issues have a test (unless they really are the same or your test covers all, just indicate this in a comment)

Commit:
TST: Split test for #15912

Separate the test for #7612 and #22334, that are about string index,
and test for #15912, about multiindex.

Commit:
CLN: Fix import position in _post_plot_logic_common

Commit:
DOC: Add whatsnew entry in Plotting

nrebena commented:
I added the whatsnew entry, and splitted the test in two, one for string index and one for multiindex, as this was one of the difference between the issues, even if the solution is the same.

nrebena reviewed (commented):
Splitted the test in two, and moved the import.

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -371,7 +371,7 @@ Plotting
 - Fixed bug where :class:`api.extensions.ExtensionArray` could not be used in matplotlib plotting (:issue:`25587`)
 - Bug in an error message in :meth:`DataFrame.plot`. Improved the error message if non-numerics are passed to :meth:`DataFrame.plot` (:issue:`25481`)
--
+- If the index of a data frame is objects (i.e. non-numerical/non-datetime), the ticklabel position are now fixed using matplotlib.ticker.FixedLocator (:issue:`7612` :issue:`15912` :issue:`22334`)
```
Bug in incorrect ticklabel positions when plotting an index that are non-numeric / non-datetime

nrebena replied:
Changed as requested

jreback reviewed (changes requested):
minor doc comment; ping on green.

nrebena committed:
DOC: Minor doc correction in whatsnew entry

jreback commented:
does this fix  #11465 ? (which is marked as a duplicate); can you add a test for this one as well.

nrebena commented:
I already looked into it, and this doesnt fix #11465, and neither #2980. These are not duplicate and should maybe be reopen.
The way bar plot handle tick position is very different from how it is done in line plot, and it deal with ticks in its own peculiar way. I think I will look into them for my next PR hopefully :smiley:

jreback commented:
thanks @nrebena

jreback commented:
welcome patches for other issues!

ImportanceOfBeingErnest commented on pandas/plotting/_core.py:
```diff
@@ -422,8 +427,11 @@ def get_label(i):
+                yticklabels = [get_label(y) for y in yticks]
                 ax.set_yticklabels(yticklabels)
+                ax.xaxis.set_major_locator(FixedLocator(yticks))
+                ax.xaxis.set_major_formatter(FixedFormatter(yticklabels))
```
Why is the **x** axis given the **y**-tick(label)s here? Seems like an error to me. 
Meaning the "horizontal" direction is not sufficiently tested.

Also note that `ax.set_xticklabels(xticklabels)` and `ax.xaxis.set_major_formatter(FixedFormatter(xticklabels))` are redundant. One of those lines should be enough.

nrebena replied:
This is definitly an error. 
And this part is definitly not tested at all. None of the tests in plotting even hit this lines. This is because the self._need_to_set_index is only set to True for plot that have a self.orientation "vertical", so there is no way to go inside this loop.
What is the best course of action? Deleting this part that could never get hits?

WillAyd replied:
If you have identified dead code you can submit a follow up PR to remove
